File: app/src/main/java/com/bas/google_book_app/ui/detail/DetailViewModel.java
package com.bas.google_book_app.ui.detail;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.ViewModel;

import com.bas.google_book_app.db.BookEntry;
import com.bas.google_book_app.repository.BookRepository;
import com.bas.google_book_app.utilsdata.AppThreadExecutors;

public class DetailViewModel extends ViewModel {
    private final BookRepository mRepository;
    private LiveData<BookEntry> mBookEntry;


    public DetailViewModel(BookRepository repository, String bookId) {
        mRepository = repository;
        mBookEntry = mRepository.getFavoriteBookById(bookId);
    }

    public LiveData<BookEntry> getBookEntry() {
        return mBookEntry;
    }

    public void addFavoriteBook(BookEntry book) {
        AppThreadExecutors.getInstance().diskIO().execute(() -> mRepository.addFavoriteBook(book));
    }

    public void removeFavoriteBook(BookEntry book) {
        AppThreadExecutors.getInstance().diskIO().execute(() -> mRepository.removeFavoriteBook(book));
    }
}

File: app/src/main/java/com/bas/google_book_app/ui/detail/DetailViewModel.kt
package com.bas.google_book_app.ui.detail

import androidx.lifecycle.LiveData
import androidx.lifecycle.ViewModel
import com.bas.google_book_app.db.BookEntry
import com.bas.google_book_app.repository.BookRepository
import com.bas.google_book_app.utilsdata.AppThreadExecutors

class DetailViewModel(private val mRepository: BookRepository?, bookId: String?) : ViewModel() {
    private val mBookEntry: LiveData<BookEntry?>?
    fun getBookEntry(): LiveData<BookEntry?>? {
        return mBookEntry
    }

    fun addFavoriteBook(book: BookEntry?) {
        AppThreadExecutors.Companion.getInstance()?.diskIO()
            ?.execute(Runnable { mRepository?.addFavoriteBook(book) })
    }

    fun removeFavoriteBook(book: BookEntry?) {
        AppThreadExecutors.Companion.getInstance()?.diskIO()
            ?.execute(Runnable { mRepository?.removeFavoriteBook(book) })
    }

    init {
        mBookEntry = mRepository?.getFavoriteBookById(bookId)
    }
}